vncviewer/win32TouchHandler.cxx

```diff
@@ -0,0 +1,369 @@
+  }
+}
+*/
+
+void win32TouchHandler::fakeMotionEvent(const GHEvent origEvent) {
+  FakeEvent ev;
+  ev.Msg = WM_MOUSEMOVE;
+  ev.wParam = 0;
+  ev.lParam = MAKELPARAM(origEvent.event_x, origEvent.event_y);
+
+  switch (origEvent.gesture) {
+    case GH_ONETAP:
+    case GH_LONGPRESS:
+    case GH_DRAG:
+      // FIXME: THIS CLASS SHOULDN'T HAVE THIS KNOWLEDGE
+      ev.wParam = MK_LBUTTON;
+      break;
+    case GH_THREETAP:
+      ev.wParam = MK_MBUTTON;
+      break;
+    case GH_TWOTAP:
+      ev.wParam = MK_RBUTTON;
+      break;
+    case GH_PINCH:
+      ev.wParam = MK_CONTROL;
+      break;
+  }
+
+  vlog.error("queing mousemove (%i, %i)",
+             (int)origEvent.event_x, (int)origEvent.event_y);
+  pushFakeEvent(ev);
+}
+
+void win32TouchHandler::fakeButtonEvent(bool press, int button,
+                                        const GHEvent origEvent) {
+  FakeEvent ev;
+  ev.Msg = 0;
+  ev.lParam = 0;
+  ev.wParam = 0;
+
+  switch (button) {
+  case 1:
+    ev.Msg = press ? WM_LBUTTONDOWN : WM_LBUTTONUP;
+    ev.wParam = press ? MK_LBUTTON : 0;
+    break;
+  case 2:
+    ev.Msg = press ? WM_MBUTTONDOWN : WM_MBUTTONUP;
+    ev.wParam = press ? MK_MBUTTON : 0;
+    break;
+  case 3:
+    ev.Msg = press ? WM_RBUTTONDOWN : WM_RBUTTONUP;
+    ev.wParam = press ? MK_RBUTTON : 0;
+    break;
+  case 4:
+    ev.Msg = WM_MOUSEWHEEL;
+    ev.wParam = MAKEWPARAM(0, WHEEL_DELTA);
+    break;
+  case 5:
+    ev.Msg = WM_MOUSEWHEEL;
+    ev.wParam = MAKEWPARAM(0, -WHEEL_DELTA);
+    break;
+  case 6:
+    ev.Msg = WM_MOUSEHWHEEL;
+    ev.wParam = MAKEWPARAM(0, -WHEEL_DELTA);
+    break;
+  case 7:
+    ev.Msg = WM_MOUSEHWHEEL;
+    ev.wParam = MAKEWPARAM(0, WHEEL_DELTA);
+    break;
+  }
+
+  // FIXME: THIS CLASS SHOULDN'T HAVE THIS KNOWLEDGE
+  // Send Ctrl with PINCH events
+  if (origEvent.gesture == GH_PINCH) {
+    ev.wParam |= MK_CONTROL;
+  }
+
+  if (1 <= button && button <= 3) {
+    // Keep relative coordinates
+    ev.lParam = MAKELPARAM(origEvent.event_x, origEvent.event_y);
+  } else if (4 <= button && button <= 7) {
+    // Convert to absolute coordinates for mouse wheel events
+    POINT pos[] = { (LONG)origEvent.event_x, (LONG)origEvent.event_y };
+    ClientToScreen(hWnd, pos);
+    ev.lParam = MAKELPARAM(pos->x, pos->y);
+  }
+
+  vlog.error("queing mouse%i button%s: (0x%x)",
+             button, press ? "down" : "up", ev.Msg);
+  pushFakeEvent(ev);
+}
+
+void win32TouchHandler::fakeKeyEvent(bool press, int keysym,
+                                     const GHEvent origEvent) {
+  FakeEvent ev;
+
+  ev.Msg = press ? WM_KEYDOWN : WM_KEYUP;
+  int virtual_keycode = 0;
+  if (keysym == XK_Control_L) {
+    virtual_keycode = VK_CONTROL;
+  }
+  ev.wParam = virtual_keycode;
+  ev.lParam = virtual_keycode;
+  vlog.error("queing key%s: 0x%x", press ? "down" : "up", virtual_keycode);
+  pushFakeEvent(ev);
+}
+
+void win32TouchHandler::pushFakeEvent(FakeEvent ev) {
```

Comment:
I'm unsure if this is a good style, should I have struct for FakeEvent or not? @CendioOssman @aeneby thoughts?

Reply:
I'd avoid it.